File: app/src/main/java/com/mastip/personaltodolist/support/adapter/TaskAdapter.java
package com.mastip.personaltodolist.support.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.mastip.personaltodolist.model.GroupingTaskModel;
import com.mastip.personaltodolist.model.TaskModel;
import com.mastip.personaltodolist.R;
import com.mastip.personaltodolist.support.adapter.viewholder.JFinishViewHolder;
import com.mastip.personaltodolist.support.adapter.viewholder.JHeaderViewHolder;
import com.mastip.personaltodolist.support.adapter.viewholder.JTaskViewHolder;
import com.mastip.personaltodolist.utilities.Constant;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by HateLogcatError on 8/25/2017.
 */

public class TaskAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    ArrayList<GroupingTaskModel> groupTask = new ArrayList<>();
    ArrayList<Integer> headerPosition = new ArrayList<>();

    private Context context;

    public TaskAdapter(Context context, ArrayList<GroupingTaskModel> groupTask) {
        this.context = context;
        this.groupTask = groupTask;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = null;
        if(viewType == Constant.VIEW_HEADER) {
            v = LayoutInflater.from(context).inflate(R.layout.adapter_header_task, null);
            return new JHeaderViewHolder(v);
        }
        else if(viewType == Constant.VIEW_BODY_UNFINISH) {
            v = LayoutInflater.from(context).inflate(R.layout.adapter_task, null);
            return new JTaskViewHolder(v);
        }
        else {
            v = LayoutInflater.from(context).inflate(R.layout.adapter_task, null);
            return new JFinishViewHolder(v);
        }
    }

    @Override
    public int getItemViewType(int position) {
        int lastGroupPosition = headerPosition.get(headerPosition.size() - 1);

        int pos = position;
        for(GroupingTaskModel groupingTaskModel: groupTask) {
            if(pos == 0) {
                return Constant.VIEW_HEADER;
            }
            else if(pos <= groupingTaskModel.getTaskModel().size()) {
                if(position > lastGroupPosition && isHaveFinishData()) {
                    return Constant.VIEW_BODY_FINISH;
                }
                else {
                    return Constant.VIEW_BODY_UNFINISH;
                }
            }
            else {
                pos -= (groupingTaskModel.getTaskModel().size() + 1);
            }
        }
        return 0;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        holder.setIsRecyclable(false);

        int pos = position;
        for(GroupingTaskModel groupingTaskModel: groupTask) {
            List<TaskModel> listModel = groupingTaskModel.getTaskModel();

            if(pos == 0) {
                if(holder instanceof JHeaderViewHolder) {
                    ((JHeaderViewHolder) holder).setUpModelToUI(groupingTaskModel);
                }
                break;
            }
            else if(pos <= groupingTaskModel.getTaskModel().size()) {
                pos -= 1;

                TaskModel taskModel = listModel.get(pos);

                if(holder instanceof JTaskViewHolder) {
                    ((JTaskViewHolder) holder).setUpModelToUI(context, taskModel);
                }
                else if(holder instanceof JFinishViewHolder) {
                    ((JFinishViewHolder) holder).setUpModelToUI(context, taskModel);
                }
                break;
            }
            else {
                pos -= (groupingTaskModel.getTaskModel().size() + 1);
            }
        }
    }

    @Override
    public int getItemCount() {
        return getCountGroupData();
    }

    private int getCountGroupData() {
        headerPosition.clear();
        int positionCount = 0;

        for (GroupingTaskModel groupingTaskModel: groupTask) {
            headerPosition.add(positionCount);
            positionCount = positionCount + groupingTaskModel.getTaskModel().size() + 1;
        }

        return positionCount;
    }

    private boolean isHaveFinishData() {
        return groupTask.get(groupTask.size() - 1).getGroupName().equals(Constant.TAG_MY_TASK_FINISH);
    }
}

File: app/src/main/java/com/mastip/personaltodolist/support/adapter/TaskAdapter.kt
package com.mastip.personaltodolist.support.adapter

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.mastip.personaltodolist.R
import com.mastip.personaltodolist.model.GroupingTaskModel
import com.mastip.personaltodolist.support.adapter.viewholder.JFinishViewHolder
import com.mastip.personaltodolist.support.adapter.viewholder.JHeaderViewHolder
import com.mastip.personaltodolist.support.adapter.viewholder.JTaskViewHolder
import com.mastip.personaltodolist.utilities.Constant
import java.util.*

class TaskAdapter(private val context: Context, private val groupTask: MutableList<GroupingTaskModel>) : androidx.recyclerview.widget.RecyclerView.Adapter<androidx.recyclerview.widget.RecyclerView.ViewHolder>() {

    var headerPosition = ArrayList<Int>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        val v: View
        return when (viewType) {
            Constant.VIEW_HEADER -> {
                v = LayoutInflater.from(context).inflate(R.layout.adapter_header_task, parent, false)
                JHeaderViewHolder(v)
            }
            Constant.VIEW_BODY_UNFINISH -> {
                v = LayoutInflater.from(context).inflate(R.layout.adapter_task, parent, false)
                JTaskViewHolder(v)
            }
            else -> {
                v = LayoutInflater.from(context).inflate(R.layout.adapter_task, parent, false)
                JFinishViewHolder(v)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        val lastGroupPosition = headerPosition[headerPosition.size - 1]
        var pos = position
        for (groupingTaskModel in groupTask) {
            pos -= if (pos == 0) {
                return Constant.VIEW_HEADER
            } else if (pos <= groupingTaskModel.taskModel.size) {
                return if (position > lastGroupPosition && isHaveFinishData) {
                    Constant.VIEW_BODY_FINISH
                } else {
                    Constant.VIEW_BODY_UNFINISH
                }
            } else {
                groupingTaskModel.taskModel.size + 1
            }
        }
        return 0
    }

    override fun onBindViewHolder(holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        holder.setIsRecyclable(false)
        var pos = position
        for (groupingTaskModel in groupTask) {
            val listModel = groupingTaskModel.taskModel
            if (pos == 0) {
                if (holder is JHeaderViewHolder) {
                    holder.setUpModelToUI(groupingTaskModel)
                }
                break
            } else if (pos <= groupingTaskModel.taskModel.size) {
                pos -= 1
                val taskModel = listModel[pos]
                if (holder is JTaskViewHolder) {
                    holder.setUpModelToUI(context, taskModel)
                } else if (holder is JFinishViewHolder) {
                    holder.setUpModelToUI(context, taskModel)
                }
                break
            } else {
                pos -= groupingTaskModel.taskModel.size + 1
            }
        }
    }

    override fun getItemCount(): Int {
        return countGroupData
    }

    private val countGroupData: Int
        get() {
            headerPosition.clear()
            var positionCount = 0
            for (groupingTaskModel in groupTask) {
                headerPosition.add(positionCount)
                positionCount += groupingTaskModel.taskModel.size + 1
            }
            return positionCount
        }

    private val isHaveFinishData: Boolean
        get() = groupTask[groupTask.size - 1].groupName == Constant.TAG_MY_TASK_FINISH
}